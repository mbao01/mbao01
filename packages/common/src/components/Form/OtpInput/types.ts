import { type OTPInputProps } from "react-otp-input";
import { InputProps } from "../Input/types";

export type OtpInputProps = {
  className?: string;
  inputProps?: Omit<InputProps, "type">;
} & Omit<OTPInputProps, "renderInput" | "onChange">;
