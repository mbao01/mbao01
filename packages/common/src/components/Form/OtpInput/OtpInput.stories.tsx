import type { Meta, StoryObj } from "@storybook/react";

import { OtpInput } from "./OtpInput";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: OtpInput,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
    // backgrounds: [{ name: 'dark background', value: '#000', default: true }],
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof OtpInput>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    inputProps: {
      placeholder: "x",
    },
  },
};

export const BorderedInput: Story = {
  args: {
    inputProps: {
      outline: true,
    },
  },
};

export const PrimaryInput: Story = {
  args: {
    inputProps: {
      variant: "primary",
    },
  },
};

export const TinyInput: Story = {
  args: {
    inputProps: {
      size: "xs",
      variant: "accent",
    },
  },
};

export const DisabledInput: Story = {
  args: {
    inputProps: {
      disabled: true,
    },
  },
};
