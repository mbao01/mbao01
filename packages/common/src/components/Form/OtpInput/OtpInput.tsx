import OTPInput from "react-otp-input";
import { type OtpInputProps } from "./types";
import { Input } from "../Input";
import { cn } from "../../../utilities";
import { useState } from "react";

export const OtpInput = ({
  className,
  inputProps,
  ...props
}: OtpInputProps) => {
  const [otp, setOtp] = useState("");

  return (
    <OTPInput
      {...props}
      value={otp}
      onChange={setOtp}
      renderInput={(renderProps) => (
        <Input
          {...inputProps}
          {...renderProps}
          className={cn(
            "!w-12 !appearance-none selection:bg-base text-base-content",
            className
          )}
        />
      )}
      containerStyle={`flex justify-center items-center flex-wrap text-2xl font-bold ${
        props.renderSeparator ? "gap-1" : "gap-x-3 gap-y-2"
      }`}
    />
  );
};
