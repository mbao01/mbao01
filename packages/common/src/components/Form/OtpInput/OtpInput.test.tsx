import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { OtpInput } from "./OtpInput";

describe("OtpInput", () => {
  it("renders with the correct initial state", () => {
    const { asFragment } = render(<OtpInput />);

    [1, 2, 3, 4].forEach((index) => {
      expect(
        screen.getByLabelText(`Please enter OTP character ${index}`)
      ).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
