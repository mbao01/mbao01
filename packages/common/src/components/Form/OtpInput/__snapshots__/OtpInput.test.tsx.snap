// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OtpInput > renders with the correct initial state 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center flex-wrap text-2xl font-bold gap-x-3 gap-y-2"
    style="display: flex; align-items: center;"
  >
    <input
      aria-label="Please enter OTP character 1"
      autocomplete="off"
      class="input rounded-md transition-all duration-100 min-h-fit h-10 w-12! appearance-none! selection:bg-base text-base-content"
      maxlength="1"
      style="width: 1em; text-align: center;"
      type="text"
      value=""
    />
    <input
      aria-label="Please enter OTP character 2"
      autocomplete="off"
      class="input rounded-md transition-all duration-100 min-h-fit h-10 w-12! appearance-none! selection:bg-base text-base-content"
      maxlength="1"
      style="width: 1em; text-align: center;"
      type="text"
      value=""
    />
    <input
      aria-label="Please enter OTP character 3"
      autocomplete="off"
      class="input rounded-md transition-all duration-100 min-h-fit h-10 w-12! appearance-none! selection:bg-base text-base-content"
      maxlength="1"
      style="width: 1em; text-align: center;"
      type="text"
      value=""
    />
    <input
      aria-label="Please enter OTP character 4"
      autocomplete="off"
      class="input rounded-md transition-all duration-100 min-h-fit h-10 w-12! appearance-none! selection:bg-base text-base-content"
      maxlength="1"
      style="width: 1em; text-align: center;"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
